### ISSUE TITLE
Bug: Overriding a method with self referential generic leads to StackOverflowException

### ISSUE THREAD

**Issue by vidl** [INRIA/spoon#2160]
Given:
```
public class A {
	<T extends Enum<T>> void m6(T t) {
	}
}
public class B extends A {
	@Override
	<S extends Enum<S>> void m6(S s) {
	}
}
```
Checking if `B#m6` overrides `A#m6` leads to a `StackOverflowException`.
I could manage to fix that in `spoon.support.visitor.MethodTypingContext#isSameMethodFormalTypeParameter` but there seams to be another issue when adopting the type (called in `spoon.support.visitor.ClassTypingContext.ExecutableContext#isSameSignatureLikeScopeMethod`, line 696). 

May be it is due to the fact, that a type definition like `S extends Enum<S>` is stored in the model like this:

0: CtTypeParameter (`S extends Enum<S>`)
1: --> superClass: CtTypeReference (`Enum<S>`)
2: ----> actualTypeArguments[0]: CtTypeParameterReference (`S`)
3: ------> superType: CtTypeReference (`Enum<S>`)
4: --------> actualTypeArguments[0]: CtTypeParameterReference (`S`)
5: ----------> superType: CtTypeReference (`java.lang.Object`)

The `spoon.support.visitor.AbstractTypingContext#adaptType` method adapts a "level 0" type to a "level 2" type, which then not equals.

**Comment by surli** [comment]
Thanks for reporting this one! I'm starting to work on it on #2161

**Comment by surli** [comment]
@pvojtechovsky do you know this one was fixed by #2406 ?

**Comment by pvojtechovsky** [comment]
You added test case in #2161 `GenericsTest#testGenericsOverriding`. It was failing at the beginning. Now this test is committed on master and it pass. So the problem is fixed. I am closing this issue.
